core/Test/Tasty/Run.hs

```diff
@@ -247,89 +234,158 @@ instance Show DependencyException where
 
 instance Exception DependencyException
 
+-- | Specifies how to calculate a dependency
+data DependencySpec
+  = ExactDep (Seq TestName) (TVar Status)
+  -- ^ Dependency specified by 'TestGroup'. Note that the first field is only
+  -- there for dependency cycle detection - which can be introduced by using
+  -- 'PatternDep'.
+  | PatternDep Expr
+  -- ^ All tests matching this 'Expr' should be considered dependencies
+  deriving (Eq)
+
+instance Show DependencySpec where
+  show (PatternDep dep) = "PatternDep (" <> show dep <> ")"
+  show (ExactDep testName _) = "ExactDep (" <> show testName <> ") (<TVar>)"
+
+-- | Dependency of a test. Either it points to an exact path it depends on, or
+-- contains a pattern that should be tested against all tests in a 'TestTree'.
+data Dependency = Dependency DependencyType DependencySpec
+  deriving (Eq, Show)
+
+-- | Is given 'Dependency' a dependency that was introduced with 'After'?
+isPatternDependency :: Dependency -> Bool
+isPatternDependency (Dependency _ (PatternDep {})) = True
+isPatternDependency _ = False
+
+data TestAction dep = TestAction
+  { testPath    :: Seq TestName
+  , testAction  :: (Seq Initializer, Seq Finalizer) -> IO ()
+  , testStatus  :: TVar Status
+  , testDepends :: Seq dep
+  }
+
+-- Behaves like a combination of fmap and foldl; it applies a function to each
+-- element of a structure, passing an accumulating parameter from left to right,
+-- and returning a final value of this accumulator together with the new
+-- structure. Monadic version of 'mapAccumL'.
+mapAccumLM :: Monad m => (acc -> x -> m (acc, y)) -> acc -> [x] -> m (acc, [y])
```

Comment:
Please rename to `mapAccumM`, similar to `base-4.18` (https://gitlab.haskell.org/ghc/ghc/-/commit/4b98c5ce971b4ea6a2ca9e44f2d068088546751a).